packages/lodestar/src/chain/genesis/genesis.ts

```diff
@@ -28,16 +29,104 @@ import {
   getTemporaryBlockHeader,
   processDeposit,
 } from "@chainsafe/lodestar-beacon-state-transition";
-import {bigIntMin} from "@chainsafe/lodestar-utils";
+import {bigIntMin, ILogger} from "@chainsafe/lodestar-utils";
+import {IBeaconDb} from "../../db";
+import {IEth1Notifier, IDepositEvent} from "../../eth1";
+import pipe from "it-pipe";
+import {ethers} from "ethers";
+import {IGenesisBuilder, IGenesisBuilderModules} from "./interface";
 
+export class GenesisBuilder implements IGenesisBuilder {
+  private readonly config: IBeaconConfig;
+  private readonly db: IBeaconDb;
+  private readonly eth1: IEth1Notifier;
+  private readonly logger: ILogger;
+  private state: TreeBacked<BeaconState>;
+  private depositTree: TreeBacked<List<Root>>;
+  private deposits: Deposit[];
+
+  constructor(config: IBeaconConfig, {eth1, db, logger}: IGenesisBuilderModules) {
+    this.state = getGenesisBeaconState(
+      config,
+      config.types.Eth1Data.defaultValue(),
+      getTemporaryBlockHeader(
+        config,
+        getEmptyBlock()
+      )
+    );
+    this.depositTree = config.types.DepositDataRootList.tree.defaultValue();
+    this.config = config;
+    this.eth1 = eth1;
+    this.db = db;
+    this.logger = logger;
+    this.deposits = [];
+  }
+
+  /**
+   * Get eth1 deposit events and blocks and apply to this.state until we found genesis.
+   */
+  public waitForGenesis = async (): Promise<TreeBacked<BeaconState>> => {
+    await this.initialize();
+    const eth1DataStream = await this.eth1.startProcessEth1Blocks(true);
+    return await pipe(eth1DataStream,
+      async (source: AsyncIterable<[IDepositEvent[], ethers.providers.Block]>) => {
```

Comment:
can we move it to separate function `processDepositEvents` so we can test that independently and it would help with readability